library/base64.c

```diff
@@ -267,26 +165,56 @@ int mbedtls_base64_encode( unsigned char *dst, size_t dlen, size_t *olen,
     return( 0 );
 }
 
+/* Given a Base64 digit, return its value.
+ * If c is not a Base64 digit ('A'..'Z', 'a'..'z', '0'..'9', '+' or '/'),
+ * return -1.
+ *
+ * The implementation assumes that letters are consecutive (e.g. ASCII
+ * but not EBCDIC).
+ *
+ * The implementation is constant-flow (no branch or memory access depending
+ * on the value of c) unless the compiler inlines and optimizes a specific
+ * access.
+ */
+static signed char dec_value( unsigned char c )
+{
+    unsigned char val = 0;
+    /* For each range of digits, if c is in that range, mask val with
+     * the corresponding value. Since c can only be in a single range,
+     * only at most one masking will change val. Set val to one plus
+     * the desired value so that it stays 0 if c is in none of the ranges. */
+    val |= mask_of_range( 'A', 'Z', c ) & ( c - 'A' +  0 + 1 );
+    val |= mask_of_range( 'a', 'z', c ) & ( c - 'a' + 26 + 1 );
+    val |= mask_of_range( '0', '9', c ) & ( c - '0' + 52 + 1 );
+    val |= mask_of_range( '+', '+', c ) & ( c - '+' + 62 + 1 );
```

Comment:
Note: I think there's an optimization-readability trade-off here and on the following line: leaving out `c - '+'` would probably result in slightly smaller and faster code, but make the pattern less regular when reading the code. Not sure if the difference is significant or even measurable though (I'll have a try after finishing my review).

Reply:
Ok, I tried compiling for a Cortex-M0 CPU with arm-none-eabi-gcc 9.3.1 with `-Os` and it turns out boths versions have the same code size, though they do generate different code - if anything, the code without `c - '+'` looks a bit worse as it seems to be pushing more on the stack.

On the performance front, on my laptop, the difference appears to be measurable but not very significant: running the test program with 100 iterations a couple of times give results in the range 466-471 ms with your version, and 448-453 ms without the `c - '+'` and `c - '/'`, so a relative difference of about 4%. Not sure it matters.

Reply:
With arm-none-eabi-gcc 7.3.1 I'm getting 2 more instructions in the thumb build with constants, and the timing variations are well within the variance from one execution to the next. Given that the differences are minute and inconsistent, I'll keep the regular pattern.